pkg/drainer/drainer.go

```diff
@@ -26,74 +26,97 @@ import (
-		err = drain.RunCordonOrUncordon(drainHelper, node, true)
-		if err != nil {
-			log.Fatalf("Couldn't cordon node %q: %s\n", node, err.Error())
-		}
+		return drainHelper, nil
 	}
 
-	if nthConfig.DryRun {
-		log.Printf("Node %s would have been drained, but dry-run flag was set", nthConfig.NodeName)
-	} else {
-		// Delete all pods on the node
-		err = drain.RunNodeDrain(drainHelper, nthConfig.NodeName)
-		if err != nil {
-			log.Fatalln(err.Error())
-		}
+	clusterConfig, err := rest.InClusterConfig()
+	if err != nil {
+		log.Println("Failed to create in-cluster config: ", err)
+		return &drain.Helper{}, err
 	}
+
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(clusterConfig)
+	if err != nil {
+		log.Println("Failed to create kubernetes clientset: ", err)
+		return &drain.Helper{}, err
+	}
+	drainHelper.Client = clientset
+
+	return drainHelper, nil
 }
 
-func getDrainHelper(nthConfig config.Config) *drain.Helper {
-	var clientset = &kubernetes.Clientset{}
-	if !nthConfig.DryRun {
-		config, err := rest.InClusterConfig()
-		if err != nil {
-			log.Fatalln("Failed to create in-cluster config: ", err.Error())
-		}
+// FetchNode will send an http request to the k8s api server and return the corev1 model node
+func (d Drainer) FetchNode() (*corev1.Node, error) {
+	node := &corev1.Node{}
+	if d.nthConfig.DryRun {
+		return node, nil
+	}
+	return d.drainHelper.Client.CoreV1().Nodes().Get(d.nthConfig.NodeName, metav1.GetOptions{})
+}
 
-		// creates the clientset
-		clientset, err = kubernetes.NewForConfig(config)
-		if err != nil {
-			log.Fatalln("Failed to create kubernetes clientset: ", err.Error())
-		}
+//Drain will cordon the node and evict pods based on the config
+func (d Drainer) Drain() error {
+	if d.nthConfig.DryRun {
+		log.Printf("Node %s would have been cordoned and drained, but dry-run flag was set\n", d.nthConfig.NodeName)
+		return nil
+	}
+	err := d.cordonNode()
+	if err != nil {
+		log.Println("There was an error in the drain process while cordoning the node: ", err)
+		return err
 	}
+	// Delete all pods on the node
+	err = drain.RunNodeDrain(d.drainHelper, d.nthConfig.NodeName)
+	if err != nil {
+		log.Println("There was an error in the drain process while evicting pods: ", err)
+		return err
+	}
+	return nil
+}
 
-	return &drain.Helper{
-		Client:              clientset,
-		Force:               true,
-		GracePeriodSeconds:  nthConfig.PodTerminationGracePeriod,
-		IgnoreAllDaemonSets: nthConfig.IgnoreDaemonSets,
-		DeleteLocalData:     nthConfig.DeleteLocalData,
-		Timeout:             time.Duration(nthConfig.NodeTerminationGracePeriod) * time.Second,
-		Out:                 os.Stdout,
-		ErrOut:              os.Stderr,
+// Cordon will add a NoSchedule on the node
+func (d Drainer) cordonNode() error {
+	node, err := d.FetchNode()
+	if err != nil {
+		log.Println("There was an error which fetching the kubernetes node: ", err)
+		return err
+	}
+	err = drain.RunCordonOrUncordon(d.drainHelper, node, true)
```

Comment:
Before calling cordon, can we add an extra check to see if node is already marked unschedulable ? 

```
if node.Spec.Unschedulable {
		log.Infof("node '%s' is already unschedulable.\n", d.node)
		return nil
	}

Reply:
I'm not sure this is necessary unless it just serves as an informational message. It doesn't really do any harm in calling cordon again if the node is already cordoned. Kubernetes is just going to converge to the desired state. If it's already cordoned, there will just be less work to do to get there.

Reply:
Yes, agreed with @bwagner5 here. The whole cordon process really is just tainting the node as unscheduleable, and it's an idempotent process, so I don't believe another check here is necessary.